internal/flow/flow.go

```diff
@@ -350,3 +420,65 @@ func (s *Service) cleanCopy(ctx context.Context, src, dest string) error {
 	}
 	return nil
 }
+
+func zipFiles(files []fileInfo) ([]byte, error) {
+	buf := new(bytes.Buffer)
+	w := zip.NewWriter(buf)
+	for _, file := range files {
+		f, err := w.Create(file.relativePath)
+		if err != nil {
+			return nil, err
+		}
+		fileContent, err := ioutil.ReadFile(file.fullPath)
+		if err != nil {
+			return nil, err
+		}
+		_, err = f.Write(fileContent)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Make sure to check the error on Close.
+	err := w.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	return buf.Bytes(), nil
+}
+
+func uploadFile(url string, fileContent []byte, artifactSpec artifact.Spec) error {
+	// TODO: MD5
+	//h := md5.New()
+	//md5s := base64.StdEncoding.EncodeToString
+
+	req, err := http.NewRequest(http.MethodPut, url, bytes.NewReader(fileContent))
+	//req, err := http.NewRequest(http.MethodPut, url, strings.NewReader(""))
+	if err != nil {
+		return err
+	}
+
+	jsonSpec, err := artifact.Encode(artifactSpec, false)
+	if err != nil {
+		return err
+	}
+	req.Header.Set("x-amz-meta-artifact-spec", jsonSpec)
+
+	//req.Header.Set("Content-MD5", md5s)
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return err
+	}
+
```

Comment:
We should make sure to close the body here.

```suggestion
	defer resp.Body.Close()
```

Reply:
👍